pkg/services/dns/dns.go

```diff
@@ -111,13 +113,45 @@ func (h *dnsHandler) addAnswers(m *dns.Msg) {
 	}
 }
 
-func Serve(udpConn net.PacketConn, zones []types.Zone) error {
-	mux := dns.NewServeMux()
+type Server struct {
+	udpConn net.PacketConn
+	handler *dnsHandler
+}
+
+func New(udpConn net.PacketConn, zones []types.Zone) (*Server, error) {
 	handler := &dnsHandler{zones: zones}
-	mux.HandleFunc(".", handler.handle)
+	return &Server{udpConn: udpConn, handler: handler}, nil
+}
+
+func (s *Server) Serve() error {
+	mux := dns.NewServeMux()
+	mux.HandleFunc(".", s.handler.handle)
 	srv := &dns.Server{
-		PacketConn: udpConn,
+		PacketConn: s.udpConn,
 		Handler:    mux,
 	}
 	return srv.ActivateAndServe()
 }
+
+func (s *Server) Mux() http.Handler {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/all", func(w http.ResponseWriter, r *http.Request) {
+		_ = json.NewEncoder(w).Encode(s.handler.zones)
+	})
+
+	mux.HandleFunc("/add", func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodPost {
+			http.Error(w, "post only", http.StatusBadRequest)
+			return
+		}
+		var req types.Zone
+		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		s.handler.zones = append([]types.Zone{req}, s.handler.zones...)
```

Comment:
After this change, we'll have multiple go routines reading or writing to `s.handler.zones`, and these accesses can be concurrent, it needs to be protected with a `sync.Mutex` or `sync.RWMutex`. If you follow the other suggestion I made (splitting this PR in 2 commits), this change can go in its own commit as well, preferably before the addition of the dynamic DNS entries.

Reply:
Won't using sync.Mutex take a performance hit on every other call ?

I thought of this case, but thought add of DNS will be less so didn't do. I will give a try with sync.Mutex and do a perf check as well

Reply:
`sync.Mutex` has a cost yes, but I don't think we have a choice here. If you run `/dns/add` in a thread as fast as possible and `/dns/all` in another thread also as fast as possible, bad things are likely to happen. `go test -race` can help to observe this.

Reply:
Done. Added sync.RWMutex to allow multiple read and block read only when write lock is obtained so hopefully should not impact a lot